File: src/main/Java/boj_step21/boj24060.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj24060 {
    static int[] tmp = new int[500_000];
    static int count = 0;
    static int result = -1;
    static int temp = 0;
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        int n = Integer.parseInt(tokenizer.nextToken());
        int k = Integer.parseInt(tokenizer.nextToken());
        temp = k;
        int[] list = Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        mergeSort(list, 0, n - 1);
        writer.append(String.valueOf(result)).close();
    }

    private static void mergeSort(int[] list, int p, int r) {
        if (p < r) {
            int q = (p + r) / 2;
            mergeSort(list, p, q);
            mergeSort(list , q + 1, r);
            merge(list, p, q, r);
        }
    }
    private static void merge(int[] list, int p, int q, int r) {
        int i = p;
        int j = q + 1;
        int t = 0;
        while (i <= q && j <= r) {
            if (list[i] <= list[j]) tmp[t++] = list[i++];
            else tmp[t++] = list[j++];
        }
        while (i <= q) tmp[t++] = list[i++];
        while (j <= r) tmp[t++] = list[j++];
        i = p;
        t = 0;
        while (i <= r) {
            list[i++] = tmp[t++];
            if (++count == temp) result = tmp[t - 1];
        }
    }
}

File: src/main/Kotlin/boj_step21/boj24060.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private val tmp = IntArray(500_000)
private var count = 0
private var result = -1
private var temp = 0
fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val (n, k) = r.readLine().split(" ").map { it.toInt() }
    temp = k
    val list = r.readLine().split(" ").map { it.toInt() }.toIntArray()
    mergeSort(list, 0, n - 1)
    w.append(result.toString())
    w.close()
}

private fun mergeSort(list: IntArray, p: Int, r: Int) {
    if (p < r) {
        val q = (p + r) / 2
        mergeSort(list, p, q)
        mergeSort(list, q + 1, r)
        merge(list, p, q, r)
    }
}

private fun merge(list: IntArray, p: Int, q: Int, r: Int) {
    var i = p
    var j = q + 1
    var t = 0
    while (i <= q && j <= r) {
        if (list[i] <= list[j]) tmp[t++] = list[i++]
        else tmp[t++] = list[j++]
    }
    while (i <= q) tmp[t++] = list[i++]
    while (j <= r) tmp[t++] = list[j++]
    i = p
    t = 0
    while (i <= r) {
        list[i++] = tmp[t++]
        if (++count == temp) result = tmp[t - 1]
    }
}